server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/main/java/org/apache/james/webadmin/vault/routes/DeletedMessageVaultRestoreTask.java

```diff
@@ -0,0 +1,131 @@
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.webadmin.vault.routes;
+
+import java.util.Date;
+import java.util.Optional;
+
+import javax.mail.Flags;
+
+import org.apache.james.core.User;
+import org.apache.james.mailbox.MailboxManager;
+import org.apache.james.mailbox.MailboxSession;
+import org.apache.james.mailbox.MessageManager;
+import org.apache.james.mailbox.exception.MailboxException;
+import org.apache.james.mailbox.model.MailboxPath;
+import org.apache.james.task.Task;
+import org.apache.james.task.TaskExecutionDetails;
+import org.apache.james.vault.DeletedMessage;
+import org.apache.james.vault.DeletedMessageVault;
+import org.apache.james.vault.search.Query;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.fge.lambdas.Throwing;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+class DeletedMessageVaultRestoreTask implements Task {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DeletedMessageVaultRestoreTask.class);
+
+    private static final String TYPE = "deletedMessages/restore";
+    private static final String RESTORE_MAILBOX_NAME = "deleted messages restore";
+
+    private final DeletedMessageVault deletedMessageVault;
+    private final User userToRestore;
+    private final MailboxManager mailboxManager;
+    private final MailboxPath restoreMailbox;
+    private final MailboxSession session;
+
+    DeletedMessageVaultRestoreTask(DeletedMessageVault deletedMessageVault, User userToRestore, MailboxManager mailboxManager) {
+        this.deletedMessageVault = deletedMessageVault;
+        this.userToRestore = userToRestore;
+        this.mailboxManager = mailboxManager;
+        this.restoreMailbox = MailboxPath.forUser(userToRestore.asString(), RESTORE_MAILBOX_NAME);
+        this.session = createSession();
+    }
+
+    private MailboxSession createSession() {
+        try {
+            return mailboxManager.createSystemSession(userToRestore.asString());
+        } catch (MailboxException e) {
+            LOGGER.error("error while creating system session of user {} ", userToRestore.asString(), e);
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public Result run() {
+        MessageManager restoreMailboxManager = restoreMailboxManager();
+
+        return Flux.from(deletedMessageVault.search(userToRestore, Query.ALL)).toStream()
+            .map(deletedMessage -> moveToMailbox(restoreMailboxManager, deletedMessage))
+            .reduce(Result.COMPLETED, Task::combine);
+    }
+
+    private Result moveToMailbox(MessageManager restoreMailboxManager, DeletedMessage deletedMessage) {
+        Mono.from(deletedMessageVault.delete(userToRestore, deletedMessage.getMessageId())).block();
```

Comment:
I was thinking about the case, if I restore first, message is appeared in the user mailbox. But the second step of removing message invault doesn't success. Then I have to remove message from user mailbox to follow data consistency. Which I find is not convenience to append a message from user mailbox, then delete it. 

BTW, if appending message to user mailbox doesn't success, I had wrote a rollback statement to restore message to the vault to avoid data lost.

Reply:
Or... We have the message both in the vault AND in the user boxes. Which is not ideal but very allright.

Let's accept that rather than writting too complex code to handle these kind of acceptable corner cases.